drivers/ethernet/eth_native_posix.c

```diff
@@ -231,55 +231,134 @@ static inline struct net_if *get_iface(struct eth_context *ctx,
 #endif
 }
 
-static int read_data(struct eth_context *ctx, int fd)
+#if defined(CONFIG_NET_VLAN)
+static struct net_pkt *prepare_vlan_pkt(struct eth_context *ctx,
+					int count, u16_t *vlan_tag, int *status)
 {
-	u16_t vlan_tag = NET_VLAN_TAG_UNSPEC;
-	struct net_if *iface;
+	struct net_eth_vlan_hdr *hdr = (struct net_eth_vlan_hdr *)ctx->recv;
 	struct net_pkt *pkt;
-	int count;
 
-	count = eth_read_data(fd, ctx->recv, sizeof(ctx->recv));
-	if (count <= 0) {
-		return 0;
+	if (IS_ENABLED(CONFIG_ETH_NATIVE_POSIX_VLAN_TAG_STRIP)) {
+		count -= NET_ETH_VLAN_HDR_SIZE;
 	}
 
 	pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, count,
 					   AF_UNSPEC, 0, NET_BUF_TIMEOUT);
 	if (!pkt) {
-		return -ENOMEM;
+		*status = -ENOMEM;
+		return NULL;
+	}
+
+	net_pkt_set_vlan_tci(pkt, ntohs(hdr->vlan.tci));
+	*vlan_tag = net_pkt_vlan_tag(pkt);
+
+	if (IS_ENABLED(CONFIG_ETH_NATIVE_POSIX_VLAN_TAG_STRIP)) {
+		if (net_pkt_write(pkt, ctx->recv,
+				  2 * sizeof(struct net_eth_addr))) {
+			goto error;
+		}
+
+		if (net_pkt_write(pkt,
+				  ctx->recv + (2 * sizeof(struct net_eth_addr))
+				  + NET_ETH_VLAN_HDR_SIZE,
+				  count - (2 * sizeof(struct net_eth_addr)))) {
+			goto error;
+		}
+	} else {
```

Comment:
It's not exactly what I proposed. You could avoid that else if you had a "pos" variable set to 0 by default and have only 2 net_pkt_write in this code. Please check my very first comment again

Reply:
ok

Reply:
Updated, hope it's fine now.